src/pybind/mgr/prometheus/module.py

```diff
@@ -505,6 +508,146 @@ def get_num_objects(self):
+            'write_latency' : {'type' : self.PERFCOUNTER_LONGRUNAVG,
+                               'desc' : 'RBD image writes latency (msec)'},
+            'read_latency' : {'type' : self.PERFCOUNTER_LONGRUNAVG,
+                              'desc' : 'RBD image reads latency (msec)'},
+        }
+
+        pools = self.get_localized_config('rbd_stats_pools', '').split()
+
+        for pool_id, pool in self.rbd_stats['pools'].items():
+            if pool['name'] not in pools:
+                del self.rbd_stats['pools'][pool_id]
+
+        for pool_name in pools:
+            try:
+                pool_id = self.rados.pool_lookup(pool_name)
+                with self.rados.open_ioctx(pool_name) as ioctx:
+                    if pool_id not in self.rbd_stats['pools']:
+                        self.rbd_stats['pools'][pool_id] = {'images' : {}}
+                    pool = self.rbd_stats['pools'][pool_id]
+                    pool['name'] = pool_name
+                    images = {}
+                    for image_meta in RBD().list2(ioctx):
+                        image = {'n' : image_meta['name']}
+                        image_id = image_meta['id']
+                        if image_id in pool['images']:
+                            image['c'] = pool['images'][image_id]['c']
+                        else:
+                            image['c'] = \
+                                [[0, 0] for x in COUNTER_DESCRIPTORS.keys()]
+                        images[image_id] = image
+                    pool['images'] = images
+            except Exception as e:
+                self.log.error('failed listing pool %s: %s' % (pool_name, e))
+
+        pool_ids = self.rbd_stats['pools'].keys()
+        pool_ids.sort()
+        pool_id_regex = '|'.join(['^%s$' % x for x in pool_ids])
+
+        if 'query' in self.rbd_stats and \
+           pool_id_regex != self.rbd_stats['query']['key_descriptor'][0]['regex']:
+            self.remove_osd_perf_query(self.rbd_stats['query_id'])
+            del self.rbd_stats['query_id']
+            del self.rbd_stats['query']
+
+        if not self.rbd_stats['pools']:
+            return
+
+        if 'query_id' not in self.rbd_stats:
+            query = {
+                'key_descriptor': [
+                    {'type': 'pool_id', 'regex': pool_id_regex},
+                    {'type': 'object_name',
+                     'regex': '^(?:rbd|journal)_data\.(?:([0-9]+)\.)?([^.]+)\.'},
+                ],
+                'performance_counter_descriptors': COUNTER_DESCRIPTORS.keys(),
+            }
+            query_id = self.add_osd_perf_query(query)
+            if query_id is None:
+                self.log.error('failed to add query %s' % query)
+                return
+            self.rbd_stats['query'] = query
+            self.rbd_stats['query_id'] =query_id
+
+        res = self.get_osd_perf_counters(self.rbd_stats['query_id'])
+        for c in res['counters']:
+            if c['k'][1][1]:
```

Comment:
I would have thought that the pool id would always be in ```c['k'][0][0]``` since it's the first listed sub-key above? Of course, if that's the case, I guess I don't understand why the image id would be in ```c['k'][1][2]``` below.

Reply:
Note, the regex is `'^(?:rbd|journal)_data\.(?:([0-9]+)\.)?([^.]+)\.'`. Consider an object "rbd_data.3.imageid.000xyz". The object will be in the data pool, so the pool id from "pool id" subkey (`c['k'][0][0]`) will match the data pool id, but to construct "pool/image" spec we need the metadata pool id, which can be extracted from the object name and is 3 for this example.

BTW for osd_perf_query module we have a simplified rbd query regex, which does not work for data pool and journal objects. I have a patch for it, but have not pushed it yet as it depends on "limit" PR to be  merged first.

Reply:
Ah -- missed the pool id optional in the reg ex. Can you add one-line comment here to clarify the intention?